### PR TITLE
Pagination, Closes #24, Connected to #11

### LINKED ISSUE #24: Enable pagination in POP

### PR DESCRIPTION
**Pull request by lauraluiz**
@schleichardt @peter-gerhard please review

### CODE REVIEW

File: product-catalog/app/productcatalog/pages/PaginationDataFactory.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        this.totalPages = totalPages;
+        this.displayedPages = displayedPages;
+        this.pageThresholdLeft = displayedPages - 1;
+        this.pageThresholdRight = totalPages - displayedPages + 2;
+    }
+
+    public PaginationData create() {
+        final PaginationData paginationData = new PaginationData(productsCount, totalProducts);
+
+        final List<LinkData> pages;
```
Comment (peter-gerhard) [inline product-catalog/app/productcatalog/pages/PaginationDataFactory.java]: imho, this is a bit un easy to grasp
what about sth like this: 
```
 public PaginationData create() {
        final PaginationData paginationData = new PaginationData(productsCount, totalProducts);
        final Range<Integer> pageRange = getPageRange();
        paginationData.setPages(getPages(pageRange.from, pageRange.to));
        paginationData.setFirstPage(getLinkData(pageRange.from));
        paginationData.setLastPage(getLinkData(pageRange.to));
        setPrevPage(paginationData);
        setNextPage(paginationData);
        return paginationData;
    }

    private Range<Integer> getPageRange() {
        if (totalPages <= displayedPages) {
            return new Range<>(1, totalPages);
        } else if (currentPage < pageThresholdLeft) {
            return new Range<>(1, pageThresholdLeft);
        } else if (currentPage > pageThresholdRight) {
            return new Range<>(pageThresholdRight, totalPages);
        } else {
            return new Range<>(currentPage - 1, currentPage + 1);
        }
    }
```

Comment (lauraluiz) [inline product-catalog/app/productcatalog/pages/PaginationDataFactory.java]: Yes, it is annoying and I've tried improving it, but the First and List pages mess with the code.
Your idea is good, but First and Last are always set, while in my code they are only set in certain situations.
I'll try to work a bit more on it after your idea.

File: common/app/common/utils/UrlUtils.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public static String buildQueryString(final Map<String, String[]> queryString) {
+        return queryString.entrySet().stream()
+                .map(parameter -> buildQueryStringOfParameter(parameter.getKey(), parameter.getValue()))
+                .collect(joining("&"));
+    }
+
+    private static String buildQueryStringOfParameter(final String key, final String[] values) {
+        return Stream.of(values).collect(joining("&" + key + "=", key + "=", ""));
```
Comment (schleichardt) [inline common/app/common/utils/UrlUtils.java]: interesting

**Comment by schleichardt** [comment]
:+1:

**Comment by peter-gerhard** [comment]
:+1: for me as well

File: product-catalog/app/productcatalog/pages/PaginationDataFactory.java (hunk 1 above)
Comment (lauraluiz) [inline product-catalog/app/productcatalog/pages/PaginationDataFactory.java]: Didn't manage to improve it now, but I'll totally work on this in the future.

**Comment by lauraluiz** [comment]
I added some tests and a little improvement. I'll merge it anyway since the tests won't affect the functionality. You can still review these 2 commits any time and I'll modify them if necessary @peter-gerhard @schleichardt

**Comment by lauraluiz** [comment]
@peter-gerhard the SelectableLinkData is just an extension of the LinkData, which is needed in some cases where you have a link that you might need to show as selected or not. For example in pagination, you have a list of links (each page), but one must be selected (the current page).
It can also be useful for breadcrumbs (each category/product is a link, but there is the current product/category selected).

**Comment by peter-gerhard** [comment]
@lauraluiz for the breadcrumbs it will always be the last link that is the 'active one'. 
For the pagination it is always the current page, there might be a good way to connect that in the template. At least the information is already present.
I dont really like the mixup. either something is a link and i can follow it or it is an item that can be selected.

**Pull request merged**